File: src/extraction/validation.py
from dataclasses import dataclass
from datetime import date, timedelta
from typing import Optional
import re


@dataclass
class ValidationAlert:
    severity: str  # "critical", "warning", "info"
    field: str
    message: str


def validate_romanian_cnp(cnp: str | None) -> tuple[bool, str]:
    """Validates length, character set, and the MOD-11 control digit of a Romanian CNP."""
    if not cnp:
        return False, "CNP is missing."

    clean_cnp = re.sub(r"\s+", "", cnp).strip()
    if len(clean_cnp) != 13 or not clean_cnp.isdigit():
        return False, f"CNP '{clean_cnp}' must be exactly 13 digits."

    # Standard Romanian CNP validation key
    CONTROL_KEY = [2, 7, 9, 1, 4, 6, 3, 5, 8, 2, 7, 9]
    digits = [int(d) for d in clean_cnp]

    checksum = sum(d * k for d, k in zip(digits[:12], CONTROL_KEY)) % 11
    expected_control = 1 if checksum == 10 else checksum

    if digits[12] != expected_control:
        return False, f"CNP '{clean_cnp}' failed mathematical checksum verification."

    return True, "CNP is valid."


def normalize_name_tokens(name: str | None) -> set[str]:
    """Strips whitespace, punctuation, and lowercases tokens for order-agnostic comparison."""
    if not name:
        return set()
    cleaned = re.sub(r"[^a-zA-Z\s\-]", "", name.lower())
    return {token for token in cleaned.replace("-", " ").split() if len(token) > 1}


def validate_cross_documents(
    contract_data: dict,
    national_id_data: dict,
    medical_data: dict,
) -> list[ValidationAlert]:
    """Cross-references extracted entities across all scanned documents."""
    alerts: list[ValidationAlert] = []

    c_cnp = (contract_data.get("national_id") or "").strip()
    id_cnp = (national_id_data.get("national_id") or "").strip()

    # 1. CNP Discrepancy & Checksum
    if id_cnp:
        is_valid, reason = validate_romanian_cnp(id_cnp)
        if not is_valid:
            alerts.append(ValidationAlert("critical", "national_id", f"National ID Card: {reason}"))

    if c_cnp and id_cnp:
        if c_cnp != id_cnp:
            alerts.append(
                ValidationAlert(
                    "critical",
                    "national_id",
                    f"CRITICAL DISCREPANCY: CNP on Contract ('{c_cnp}') does NOT match National ID ('{id_cnp}').",
                )
            )

    # 2. Name Consistency (Contract vs National ID)
    c_name_tokens = normalize_name_tokens(f"{contract_data.get('first_name', '')} {contract_data.get('last_name', '')}")
    id_name_tokens = normalize_name_tokens(f"{national_id_data.get('first_name', '')} {national_id_data.get('last_name', '')}")

    if c_name_tokens and id_name_tokens:
        if not c_name_tokens.intersection(id_name_tokens):
            alerts.append(
                ValidationAlert(
                    "critical",
                    "name",
                    f"Name mismatch: Contract name ('{contract_data.get('first_name')} {contract_data.get('last_name')}') "
                    f"does not match National ID ('{national_id_data.get('first_name')} {national_id_data.get('last_name')}').",
                )
            )
        elif c_name_tokens != id_name_tokens:
            alerts.append(
                ValidationAlert(
                    "warning",
                    "name",
                    "Minor name variation detected between Contract and National ID (e.g., middle name omitted).",
                )
            )

    # 3. Medical Clearance Gatekeeper
    med_status = medical_data.get("medical_clearance_status")
    if med_status is False:
        alerts.append(
            ValidationAlert(
                "critical",
                "medical_clearance",
                "COMPLIANCE GATE: Medical certificate indicates candidate is UNFIT or clearance is unconfirmed.",
            )
        )

    # 4. Temporal Sanity (Medical Exam vs Contract Start Date)
    exam_date = medical_data.get("issue_date")
    start_date = contract_data.get("start_date")

    if exam_date and start_date:
        if isinstance(exam_date, str):
            exam_date = date.fromisoformat(exam_date)
        if isinstance(start_date, str):
            start_date = date.fromisoformat(start_date)

        # Clearance older than 180 days is legally stale
        if (start_date - exam_date) > timedelta(days=180):
            alerts.append(
                ValidationAlert(
                    "warning",
                    "medical_clearance_date",
                    f"Medical clearance examination date ({exam_date}) is more than 6 months prior to start date ({start_date}).",
                )
            )
        elif exam_date > (start_date + timedelta(days=14)):
            alerts.append(
                ValidationAlert(
                    "warning",
                    "medical_clearance_date",
                    f"Medical clearance date ({exam_date}) is set well after the employment start date ({start_date}).",
                )
            )

    return alerts

File: frontend/src/utils/validation.ts
export interface ValidationAlert {
  severity: 'critical' | 'warning' | 'info';
  field: string;
  message: string;
}

export function formatName(name: string | null | undefined): string {
  if (!name || typeof name !== 'string') return '';

  const capitalizePart = (part: string): string => {
    if (!part) return '';
    if (part.toLowerCase().startsWith('mc') && part.length > 2) {
      return 'Mc' + part.charAt(2).toUpperCase() + part.slice(3).toLowerCase();
    }
    return part.charAt(0).toUpperCase() + part.slice(1).toLowerCase();
  };

  const capitalizeToken = (token: string): string => {
    return token.split("'").map(capitalizePart).join("'");
  };

  return name
    .trim()
    .split(/\s+/)
    .map((word) => word.split('-').map(capitalizeToken).join('-'))
    .join(' ');
}

export function validateRomanianCnp(cnp: string | null | undefined): { isValid: boolean; reason: string } {
  if (!cnp) return { isValid: false, reason: 'CNP is missing.' };

  const clean = cnp.replace(/\s+/g, '').trim();
  if (clean.length !== 13 || !/^\d+$/.test(clean)) {
    return { isValid: false, reason: `CNP '${clean}' must be exactly 13 digits.` };
  }

  const CONTROL_KEY = [2, 7, 9, 1, 4, 6, 3, 5, 8, 2, 7, 9];
  const digits = clean.split('').map(Number);

  const checksum = digits.slice(0, 12).reduce((acc, digit, idx) => acc + digit * CONTROL_KEY[idx], 0) % 11;
  const expectedControl = checksum === 10 ? 1 : checksum;

  if (digits[12] !== expectedControl) {
    return { isValid: false, reason: `CNP '${clean}' failed mathematical checksum verification.` };
  }

  return { isValid: true, reason: 'CNP is valid.' };
}

function normalizeNameTokens(name: string | null | undefined): Set<string> {
  if (!name) return new Set();
  const cleaned = name.toLowerCase().replace(/[^a-zA-Z\s\-]/g, '');
  const tokens = cleaned.replace(/-/g, ' ').split(/\s+/).filter((t) => t.length > 1);
  return new Set(tokens);
}

export function validateCrossDocuments(
  contractData: Record<string, any>,
  nationalIdData: Record<string, any>,
  medicalData: Record<string, any>
): ValidationAlert[] {
  const alerts: ValidationAlert[] = [];
  const cCnp = (contractData.national_id || '').trim();
  const idCnp = (nationalIdData.national_id || '').trim();

  // 1. CNP Discrepancy & Checksum
  if (idCnp) {
    const { isValid, reason } = validateRomanianCnp(idCnp);
    if (!isValid) {
      alerts.push({ severity: 'critical', field: 'national_id', message: `National ID Card: ${reason}` });
    }
  }

  if (cCnp && idCnp && cCnp !== idCnp) {
    alerts.push({
      severity: 'critical',
      field: 'national_id',
      message: `CRITICAL DISCREPANCY: CNP on Contract ('${cCnp}') does NOT match National ID ('${idCnp}').`,
    });
  }

  // 2. Name Consistency
  const cTokens = normalizeNameTokens(`${contractData.first_name || ''} ${contractData.last_name || ''}`);
  const idTokens = normalizeNameTokens(`${nationalIdData.first_name || ''} ${nationalIdData.last_name || ''}`);

  if (cTokens.size > 0 && idTokens.size > 0) {
    const intersection = [...cTokens].filter((x) => idTokens.has(x));
    if (intersection.length === 0) {
      alerts.push({
        severity: 'critical',
        field: 'name',
        message: `Name mismatch: Contract name does not match National ID.`,
      });
    } else if (cTokens.size !== idTokens.size) {
      alerts.push({
        severity: 'warning',
        field: 'name',
        message: 'Minor name variation detected between Contract and National ID (e.g., omitted middle name).',
      });
    }
  }

  // 3. Medical Clearance Gatekeeper
  if (medicalData.medical_clearance_status === false) {
    alerts.push({
      severity: 'critical',
      field: 'medical_clearance',
      message: 'COMPLIANCE GATE: Medical certificate indicates candidate is UNFIT or unconfirmed.',
    });
  }

  // 4. Temporal Sanity
  if (medicalData.issue_date && contractData.start_date) {
    const examDate = new Date(medicalData.issue_date);
    const startDate = new Date(contractData.start_date);
    const diffDays = (startDate.getTime() - examDate.getTime()) / (1000 * 3600 * 24);

    if (diffDays > 180) {
      alerts.push({
        severity: 'warning',
        field: 'medical_clearance_date',
        message: `Medical clearance is older than 6 months prior to start date.`,
      });
    } else if (diffDays < -14) {
      alerts.push({
        severity: 'warning',
        field: 'medical_clearance_date',
        message: `Medical clearance date is set well after employment start date.`,
      });
    }
  }

  return alerts;
}

export function matchOption(value: string | null | undefined, options: readonly string[], fallback: string): string {
  if (!value || !value.trim()) return fallback;
  const clean = value.trim().toLowerCase();
  const matched = options.find((opt) => opt.toLowerCase().includes(clean) || clean.includes(opt.toLowerCase()));
  return matched || fallback;
}